subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/writer/WriteDependencyVerificationFile.java

```diff
@@ -502,57 +519,65 @@ private void exportKeyRingCollection(PublicKeyService publicKeyService, BuildTre
             .filter(WriteDependencyVerificationFile::hasAtLeastOnePublicKey)
             .filter(e -> existingRings.stream().noneMatch(ring -> keyIds(ring).equals(keyIds(e))))
             .collect(Collectors.toList());
-        ImmutableList<PGPPublicKeyRing> allKeyRings = ImmutableList.<PGPPublicKeyRing>builder()
+
+        return ImmutableList.<PGPPublicKeyRing>builder()
             .addAll(existingRings)
             .addAll(keysSeenInVerifier)
             .build();
-        File keyringFile = keyrings.getBinaryKeyringsFile();
-        writeBinaryKeyringFile(keyringFile, allKeyRings);
-        File asciiArmoredFile = keyrings.getAsciiKeyringsFile();
-        writeAsciiArmoredKeyRingFile(asciiArmoredFile, allKeyRings);
-        LOGGER.lifecycle("Exported {} keys to {} and {}", allKeyRings.size(), keyringFile, asciiArmoredFile);
     }
 
-    private void writeAsciiArmoredKeyRingFile(File ascii, ImmutableList<PGPPublicKeyRing> allKeyRings) throws IOException {
-        if (ascii.exists()) {
-            ascii.delete();
-        }
-        boolean hasKey = false;
-        for (PGPPublicKeyRing keyRing : allKeyRings) {
-            // First let's write some human readable info about the keyring being serialized
+    protected static Collection<PGPPublicKey> collectDistinctPublicKeys(Collection<PGPPublicKeyRing> keyRings) {
+        // Extract all the distinct public keys from the keyrings, and deduplicate them by the keyID
+        Map<Long, PGPPublicKey> allKeys = keyRings.stream()
+            .map(PGPPublicKeyRing::getPublicKeys)
+            .flatMap(keyIter -> {
+                Spliterator<PGPPublicKey> keySpliterator = Spliterators.spliteratorUnknownSize(keyIter, Spliterator.SIZED | Spliterator.IMMUTABLE);
+                return StreamSupport.stream(keySpliterator, false);
+            })
+            .collect(Collectors.toMap(PGPPublicKey::getKeyID, Function.identity(), (a, b) -> {
+                    LOGGER.debug("Found duplicated verification key: " + a.getKeyID());
+                    return a;
+                }
+            ));
+
+        // We sort all the key entries by their ID's.
+        return allKeys.entrySet()
+            .stream()
+            .sorted(Map.Entry.comparingByKey())
+            .map(Map.Entry::getValue)
+            .collect(Collectors.toList());
```

Comment:
I came to the conclusion that this is all we can do.
An optimal solution would be to sort the keyrings first, but sorting the public keys alone should provide enough stability.